SPECS/tdnf/virtual-repo-snapshot.patch

```diff
@@ -0,0 +1,906 @@
+     __cmds="autoerase autoremove check check-local check-update clean distro-sync downgrade erase help history info install list makecache mark provides whatprovides reinstall remove repolist repoquery reposync search update update-to updateinfo upgrade upgrade-to"
+     cur="${COMP_WORDS[COMP_CWORD]}"
+     _tdnf__process_if_prev_is_option && return 0
+diff --git a/include/tdnferror.h b/include/tdnferror.h
+index c9349a0..02b8d4c 100644
+--- a/include/tdnferror.h
++++ b/include/tdnferror.h
+@@ -187,6 +187,15 @@ extern "C" {
+ #define ERROR_TDNF_HISTORY_ERROR 1801
+ #define ERROR_TDNF_HISTORY_NODB 1802
+ 
++#define ERROR_TDNF_TIME_FILTER_BASE 1900
++// filter MEMORY
++#define ERROR_TDNF_TIME_FILTER_MEMORY        (ERROR_TDNF_TIME_FILTER_BASE + 1)
++// filter parsing error
++#define ERROR_TDNF_TIME_FILTER_PARSE         (ERROR_TDNF_TIME_FILTER_BASE + 2)
++// filter IO error
++#define ERROR_TDNF_TIME_FILTER_IO            (ERROR_TDNF_TIME_FILTER_BASE + 3)
++// filter general error
++# define ERROR_TDNF_TIME_FILTER_GENERAL      (ERROR_TDNF_TIME_FILTER_BASE + 4)
+ 
+ #define ERROR_TDNF_PLUGIN_BASE          2000
+ 
+diff --git a/include/tdnftypes.h b/include/tdnftypes.h
+index a806010..e638427 100644
+--- a/include/tdnftypes.h
++++ b/include/tdnftypes.h
+@@ -9,6 +9,7 @@
+ #pragma once
+ 
+ #include <curl/curl.h>
++#include <time.h>
+ 
+ #ifdef __cplusplus
+ extern "C" {
+@@ -257,6 +258,7 @@ typedef struct _TDNF_CONF
+     int nCheckUpdateCompat;
+     int nDistroSyncReinstallChanged;
+     char* pszRepoDir;
++    char* pszRepoEpochTimeLimit;
+     char* pszCacheDir;
+     char* pszPersistDir;
+     char* pszProxy;
+diff --git a/solv/defines.h b/solv/defines.h
+index 38f5ab1..ebc7bd0 100644
+--- a/solv/defines.h
++++ b/solv/defines.h
+@@ -18,4 +18,32 @@
+         }                                                          \
+     } while(0)
+ 
++typedef struct {
++    // frequently changed values
++    char * pszElementBuffer;
++    int nBufferLen;
++    int nInPackage;
++    int nPrintPackage;
++    int nTimeFound;
++
++    // managed values
++    int nBufferMaxLen;
++    int nDepth;
++    int nPrevElement; // enum 0 -> start, 1 -> data, 2 -> end
++
++    //set and forget on creation
++    time_t nSearchTime;
++    FILE * pbOutfile;
++} TDNFFilterData;
++
++#define TDNF_MAX_FILTER_INPUT_THRESHOLD 10000000
++
++#define BAIL_ON_TDNF_TIME_FILTER_ERROR(dwError) \
```

Comment:
Is there another similar macro that does the same thing?  Do we need a new one here?

Reply:
There are similar macros... but stylistically (and in other tdnf code), adding another bail specifically for the time filter seemed appropriate (see BAIL_ON_TDNF_LIBSOLV_ERROR() for example). We could likely use `BAIL_ON_TDNF_ERROR` if that seems more appropriate.